### PR TITLE
Fixed #26500 -- Added SKIP LOCKED support to select_for_update().

### PR DESCRIPTION
**Pull request by charettes**

### CODE REVIEW

**Comment by charettes** [comment]
Buildbot, test on oracle.

**Comment by charettes** [comment]
I tested the patch locally against PostgreSQL 9.5 through Docker but the tests are skipped on the CI server since it only ships with 9.4.

File: docs/ref/models/querysets.txt [hunk 1]
```diff
@@ -1534,17 +1534,22 @@ selected rows, the query will block until the lock is released. If this is
...
 another transaction, :exc:`~django.db.DatabaseError` will be raised when the
-queryset is evaluated.
+queryset is evaluated. You can also ignore locked rows by using
+``select_for_update(skip_locked=True)`` instead. Note that the ``nowait`` and
+``skip_locked`` are mutually exclusive and attempts to call
+``select_for_update()`` with both options enabled will result in a
+:exc:`ValueError`.
 
 Currently, the ``postgresql``, ``oracle``, and ``mysql`` database
 backends support ``select_for_update()``. However, MySQL has no support for the
```
Comment (timgraham) [inline docs/ref/models/querysets.txt]: doesn't support the

File: docs/ref/models/querysets.txt [hunk 2]
```diff
@@ -1534,17 +1534,22 @@ selected rows, the query will block until the lock is released. If this is
...
-``nowait`` argument. Obviously, users of external third-party backends should
-check with their backend's documentation for specifics in those cases.
+``nowait`` and ``skip_locked`` arguments. Obviously, users of external
```
Comment (timgraham) [inline docs/ref/models/querysets.txt]: I think we could chop the "Obviously," sentence.

File: docs/ref/models/querysets.txt [hunk 3]
```diff
@@ -1534,17 +1534,22 @@ selected rows, the query will block until the lock is released. If this is
 not the behavior you want, call ``select_for_update(nowait=True)``. This will
 make the call non-blocking. If a conflicting lock is already acquired by
 another transaction, :exc:`~django.db.DatabaseError` will be raised when the
-queryset is evaluated.
+queryset is evaluated. You can also ignore locked rows by using
+``select_for_update(skip_locked=True)`` instead. Note that the ``nowait`` and
```
Comment (timgraham) [inline docs/ref/models/querysets.txt]: chop "Note that"

File: docs/ref/models/querysets.txt [hunk 4]
```diff
@@ -1568,6 +1573,10 @@ raised if ``select_for_update()`` is used in autocommit mode.
     ``select_for_update()`` you should use
     :class:`~django.test.TransactionTestCase`.
 
+.. versionchanged:: 1.11
+
+    Added support for the ``skip_locked`` option.
```
Comment (timgraham) [inline docs/ref/models/querysets.txt]: The skip_locked option was added.

File: docs/releases/1.11.txt [hunk 5]
```diff
@@ -150,7 +150,8 @@ CSRF
 Database backends
 ~~~~~~~~~~~~~~~~~
 
-* ...
+* Added support for ``SKIP LOCKED`` to :meth:`QuerySet.select_for_update()
```
Comment (timgraham) [inline docs/releases/1.11.txt]: "Added the skip_locked argument to...."

For less verbosity, you can write the link as:
```
:meth:`.QuerySet.select_for_update()`
````

File: docs/releases/1.11.txt [hunk 6]
```diff
@@ -150,7 +150,8 @@ CSRF
...
+  <django.db.models.query.QuerySet.select_for_update>` on Postgres and Oracle.
```
Comment (timgraham) [inline docs/releases/1.11.txt]: PostgreSQL

File: tests/select_for_update/tests.py [hunk 7]
```diff
@@ -102,6 +109,27 @@ def test_nowait_raises_error_on_block(self):
         self.end_blocking_transaction()
         self.assertIsInstance(status[-1], DatabaseError)
 
+    @skipUnlessDBFeature('has_select_for_update_skip_locked')
+    def test_skip_locked_skips_locked_rows(self):
+        """
+        if skip_locked is specified, we expect the locked row to be skiped
```
Comment (timgraham) [inline tests/select_for_update/tests.py]: If ...., the locked row is skipped resulting in Person.DoesNotExist.

File: tests/select_for_update/tests.py [hunk 8]
```diff
@@ -102,6 +109,27 @@ def test_nowait_raises_error_on_block(self):
...
+        resulting in a `Person.DoesNotExist`.
+        """
+        self.start_blocking_transaction()
+        status = []
+
```
Comment (timgraham) [inline tests/select_for_update/tests.py]: chop blank lines

File: tests/select_for_update/tests.py [hunk 9]
```diff
@@ -113,6 +141,17 @@ def test_unsupported_nowait_raises_error(self):
         with self.assertRaises(DatabaseError):
             list(Person.objects.all().select_for_update(nowait=True))
 
+    @skipIfDBFeature('has_select_for_update_skip_locked')
+    @skipUnlessDBFeature('has_select_for_update')
+    def test_unsupported_skip_locked_raises_error(self):
+        """
+        If a SELECT...FOR UPDATE SKIP LOCKED is run on a database backend
```
Comment (timgraham) [inline tests/select_for_update/tests.py]: DatabaseError is raised if a ....

**Comment by timgraham** [comment]
Could you add a mention of `features.has_select_for_update_skip_locked` to the release notes for third-party database backends?

The first commit looks good to merge.

**Comment by timgraham** [comment]
CI server should be running PG9.5 -- did you observe otherwise?

**Comment by charettes** [comment]
That's odd, it looks like the [tests are skipped](http://djangoci.com/job/pull-requests-trusty/7918/database=postgres,label=trusty-pr,python=python2.7/testReport/select_for_update.tests/SelectForUpdateTests/test_for_update_sql_generated_skip_locked/).

File: django/db/backends/postgresql/features.py [hunk 10]
```diff
@@ -31,3 +33,7 @@ class DatabaseFeatures(BaseDatabaseFeatures):
     greatest_least_ignores_nulls = True
     can_clone_databases = True
     supports_temporal_subtraction = True
+
+    @cached_property
+    def has_select_for_update_skip_locked(self):
+        return self.connection.pg_version >= 90500
```
Comment (charettes) [inline django/db/backends/postgresql/features.py]: Can you confirm `connection.pg_version >= 90500` on the CI server?

Comment (timgraham) [inline django/db/backends/postgresql/features.py]: Should be fixed now, both 9.4 and 9.5 were installed for some reason.

**Comment by charettes** [comment]
Everything should be good now.

File: docs/releases/1.11.txt [hunk 11]
```diff
@@ -297,6 +298,9 @@ Database backend API
   support the :lookup:`time` lookup. It accepts a ``field_name`` and ``tzname``
   arguments and returns the SQL necessary to cast a datetime value to time value.
 
+* In order to enable ``FOR UPDATE SKIP LOCKED`` support the
```
Comment (timgraham) [inline docs/releases/1.11.txt]: To enable .... support, set DatabaseFeatures.has_select_for_update_skip_locked = True.

File: docs/ref/models/querysets.txt [hunk 12]
```diff
@@ -1546,17 +1546,20 @@ selected rows, the query will block until the lock is released. If this is
...
+backends support ``select_for_update()``. However, MySQL doesn't support the
+``nowait`` and ``skip_locked`` arguments.
 
-Passing ``nowait=True`` to ``select_for_update()`` using database backends that
-do not support ``nowait``, such as MySQL, will cause a
-:exc:`~django.db.DatabaseError` to be raised. This is in order to prevent code
-unexpectedly blocking.
+Passing ``nowait=True`` or ``skip_locked=True`` to ``select_for_update()``
+using database backends that do not support these options, such as MySQL, will
+cause a :exc:`~django.db.DatabaseError` to be raised. This is in order to
```
Comment (timgraham) [inline docs/ref/models/querysets.txt]: This prevents code from unexpectedly blocking.

File: tests/select_for_update/tests.py [hunk 13]
```diff
@@ -86,30 +96,56 @@ def test_nowait_raises_error_on_block(self):
         """
         self.start_blocking_transaction()
         status = []
-
```
Comment (timgraham) [inline tests/select_for_update/tests.py]: I'd rather if you made these changes including rewriting existing docstrings in a separate commit.

File: tests/select_for_update/tests.py [hunk 14]
```diff
@@ -86,30 +96,56 @@ def test_nowait_raises_error_on_block(self):
...
             list(Person.objects.all().select_for_update(nowait=True))
 
+    @skipIfDBFeature('has_select_for_update_skip_locked')
+    @skipUnlessDBFeature('has_select_for_update')
+    def test_unsupported_skip_locked_raises_error(self):
+        """
+        DatabaseError is raised if a SELECT...FOR UPDATE SKIP LOCKED is run on
+        a database backend that supports FOR UPDATE but not SKIP LOCKED.
+        """
+        with self.assertRaises(DatabaseError):
```
Comment (timgraham) [inline tests/select_for_update/tests.py]: assertRaisesMessage could be used here?

File: docs/releases/1.11.txt [hunk 15]
```diff
@@ -150,7 +150,8 @@ CSRF
 Database backends
 ~~~~~~~~~~~~~~~~~
 
-* ...
+* Added the ``skip_locked`` argument to :meth:`.QuerySet.select_for_update()`
+  on PostgreSQL and Oracle.
```
Comment (timgraham) [inline docs/releases/1.11.txt]: PostgreSQL 9.5+
add something like: "to execute queries with FOR UPDATE SKIP LOCKED"

**Pull request merged**